threads/2024/THREADS-01-09.md

```diff
@@ -0,0 +1,244 @@
+
+CW: Think of the interaction we do with JS in regular Wasm. When we import a Wasm function to JS, it's analogous to bringing a new function into JS and even in JS you’re allowed to bring a closure into cross-realm. And you expect the realm you’re referring to refer to inner realms. If you move that module to another realm, just because that module was a shared module, you have to redeclare your imports. Where you wouldn’t have to do that with non-shared. And even if in the non-shared Wasm world you’d potentially want some dynamic access…
+
+TL: Simply more consistent with the things currently working if we have thread locals instead of realm locals.
+
+RH: What's the motivation for realm locals instead of thread locals?
+
+SYG: I take the opposite opinion to CW. DX is a good thing to focus on for sure and I think for the usual case, it doesn’t matter for DX. I imagine the usual case is 1:1 of realms and threads. I imagine people are not running Sphinx in a synchronous iFrame. How I arrived at realm-local makes more sense than thread local is if you have the mental model that JS is the syscall for Wasm, then you need to hook that up somehow. Currently you look that up in a shared world by whatever thing you’re executing. In a shared world, you end up moving to a more dynamic shared place because you don’t know what thread you’re going to be running in. Technically the syscall you’re embedding is different thread to thread and realm to realm if you make the decision to be thread local, it doesn’t take away the fact that there are multiple executions on the same thread. If your JS embedding has installed polyfill, some globals inside of an iFrame and your outer realm. If you’re running some shared payload that has some TLS table setup with JS APIs. 
+
+CW: The scenario you’re describing where an individual iframe is doing a polyfill, the question of threads or not is orthogonal. What will happen in wasm is you’ll take the definition you originally imported at instantiation-time (fixing the realm) - if you’re in the single-threaded world and you transfer the Wasm to a new realm, you already have the issue of not seeing the new realm’s definition
+
+SYG: True but in a single threaded world, there is no surprise because it's a lexical scope thing. You closed over them. In the shared world, it's more dynamic. You cannot access the original context. You’re running in a new thread so where do you pick up the new context.
+
+CW: Not inconsistent to say you have a context per thread and in the same way you fix the content in the first thread by ??? you fix the content in the second by binding the secondary function.  
+
+…
+
+<TL, CW, SYG… too quick of an interchange to capture>
```

Comment:
@conrad-watt, would you be able to fill out some missing pieces (and/or condense) this part of the discussion?